### PR TITLE
[DNM] Game table

### PR DESCRIPTION
**Pull request by thoys**
## QA Test
Use the script at: https://raw.githubusercontent.com/thoys/hifi/gameTable/unpublishedScripts/marketplace/gameTable/createGameTable.js to create the gameTable.
This test requires at least two persons.
1. Press the ">>" button to setup the first game.
2. Have the other user press >> or reload, and switch around a couple of times.
3. The board picture and the pieces should match, e.g. Chessboard for Chess, Checkers board for Checkers.
4. Rotate the table in a different angle, the grid based board games should still have their pieces in the correct rotation, e.g. the chessboard knights should point towards the center.

## Dev note:
[DO NOT MERGE] this PR contains a lot of assets, we want to move this to a different repository eventually.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5c3a746f62c8cdd45d81e3226b4352e75a73e590 available at: https://gist.github.com/be11dd721a1336498bbd46ee814cc349<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-5c3a746.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-5c3a746.dmg

**Comment by Nex-Pro** [comment]
Testing this with @claudiosg

**Comment by Nex-Pro** [comment]
@thoys This was a pass for me but a fail for @claudio. 

He was using oculus and touch.  He could not interact with the buttons despite getting haptic. He tried at sandbox and in my domain to no avail.

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift

Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-5c3a746.exe

@thoys What @Nex-Pro said^^^

My avatars hands could not trigger the blue and red symbols.

**Comment by Nex-Pro** [comment]
On re-running @claudiosg was able to use the setup and reload buttons successfully.

On rotating the table and using the reload button the chess pieces correctly oriented themselves.

There appears to be an issue with the physics of the checkers pieces they fall very slowly. I do not know of this is intended. All observers saw this.

@thoys

**Comment by claudioshg** [comment]
@thoys 
 
@Nex-Pro and I were having problems testing this PR, because of my internet connection. As mentioned before, I was unable to trigger the reset and play/load buttons. This was resolved by testing in the Sandbox domain. Following the testing instructions, this PR is a PASS. Everything works as intended.

File: unpublishedScripts/marketplace/gameTable/createGameTable.js [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+var CLEARCACHE = "?"+Math.random().toString(36).substring(7);
+var TABLE_MODEL_URL = Script.resolvePath('assets/table/gameTable.fbx') + CLEARCACHE;
+var MODEL_URL = Script.resolvePath('assets/table/finalFrame.fbx') + CLEARCACHE;
+var TABLE_SCRIPT_URL = Script.resolvePath('table.js') + CLEARCACHE;
+var ENTITY_SPAWNER_SCRIPT_URL = Script.resolvePath('entitySpawner.js') + CLEARCACHE;
+var NEXT_GAME_BUTTON_SCRIPT_URL = Script.resolvePath('nextGameButton.js') + CLEARCACHE;
+var RESET_BUTTON_SCRIPT_URL = Script.resolvePath('resetGameButton.js') + CLEARCACHE;
+var TABLE_PICTURE_URL = Script.resolvePath('assets/mats/Table-default.jpg') + CLEARCACHE;
+var NEXT_BUTTON_MODEL_URL = Script.resolvePath('assets/buttons/button-next.fbx') + CLEARCACHE;
+var RESET_BUTTON_MODEL_URL = Script.resolvePath('assets/buttons/button-reset.fbx') + CLEARCACHE;
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/createGameTable.js]: Can the cache busters be removed?

File: unpublishedScripts/marketplace/gameTable/entitySpawner.js [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    function importEntitiesJSON(importLink, parentProperties, overrideProperties) {
+        if (parentProperties === undefined) {
+            parentProperties = {};
+        }
+        if (overrideProperties !== undefined) {
+            parentProperties.overrideProperties = overrideProperties;
+        }
+        var request = new XMLHttpRequest();
+        request.open('GET', importLink, false);
+        request.send();
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/entitySpawner.js]: You should be able to replace all of this with `Script.require` - a big benefit is that this should also work with atp

File: unpublishedScripts/marketplace/gameTable/entitySpawner.js [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        if (parent.id !== undefined) {
+            newEntityProperties.parentID = parent.id;
+        }
+        // Fix up paths
+        if (entityProperties.modelURL !== undefined) {
+            newEntityProperties.modelURL = fixLocalPath(newEntityProperties.modelURL);
+        }
+        if (entityProperties.script !== undefined) {
+            newEntityProperties.script = fixLocalPath(newEntityProperties.script);
+        }
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/entitySpawner.js]: I've written code like this at least a few times now. It's becoming a pattern...

File: unpublishedScripts/marketplace/gameTable/entitySpawner.js [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+                    createEntitiesFromTree(entityProperties.childEntities, parentProperties, overrideProperties);
+            }
+            createdTree.push(parentProperties);
+        });
+        return createdTree;
+    };
+
+    //listens for a release message from entities with the snap to grid script
+    //checks for the nearest snap point and sends a message back to the entity
+    function PastedItem(url, spawnLocation, spawnRotation) {
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/entitySpawner.js]: This seems like it needs a better name? Not a blocker

File: unpublishedScripts/marketplace/gameTable/nextGameButton.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        },
+        getEntityFromGroup: function(groupName, entityName) {
+            var props = Entities.getEntityProperties(_this.entityID);
+            var results = Entities.findEntities(props.position, 7.5);
+            var found = false;
+            results.forEach(function(item) {
+                var itemProps = Entities.getEntityProperties(item);
+                var descriptionSplit = itemProps.description.split(":");
+                if (descriptionSplit[1] === groupName && descriptionSplit[2] === entityName) {
+                    found = item
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/nextGameButton.js]: missing semicolon

File: unpublishedScripts/marketplace/gameTable/nextGameButton.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    NextGameButton.prototype = {
+        preload: function(id) {
+            _this.entityID = id;
+            clickSound = SoundCache.getSound(CLICK_SOUND_URL);
+            myLocation = Entities.getEntityProperties(_this.entityID).position;
+        },
+        getEntityFromGroup: function(groupName, entityName) {
+            var props = Entities.getEntityProperties(_this.entityID);
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/nextGameButton.js]: We should only request the entity properties that we will be needing - it's expensive to ask for every property if you don't need them.

File: unpublishedScripts/marketplace/gameTable/nextGameButton.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            var results = Entities.findEntities(props.position, 7.5);
+            var found = false;
+            results.forEach(function(item) {
+                var itemProps = Entities.getEntityProperties(item);
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/nextGameButton.js]: Same here - only request the properties you need

File: unpublishedScripts/marketplace/gameTable/nextGameButton.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        },
+        startNearTrigger: function() {
+            _this.nextGame();
+        },
+        startFarTrigger: function() {},
+        nextGame: function() {
+            Audio.playSound(CLICK_SOUND_URL, { loop: false, position: myLocation, volume: 0.4 });
+            var table = _this.getEntityFromGroup('gameTable', 'table');
+            var tableString = table.substr(1, table.length - 2);
+            Entities.callEntityMethod(tableString, 'nextGame');
```
Comment (huffman) [inline unpublishedScripts/marketplace/gameTable/nextGameButton.js]: Is the button a child of the game table? This is what I've done in  other situations, then you only need to look at your parentID

File: unpublishedScripts/marketplace/gameTable/entitySpawner.js (hunk 2 above)
Comment (thoys) [inline unpublishedScripts/marketplace/gameTable/entitySpawner.js]: Does the stable branch have Script.require already?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 12ef157316f3c14162c41adb1404b91e1be8d554 available at: https://gist.github.com/9a74abfb6c64450646c71046862e7b2f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-12ef157.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-12ef157.dmg

**Comment by thoys** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 588ddfde8d7455b88ef5c032f932807a95f54b3d available at: https://gist.github.com/9a74abfb6c64450646c71046862e7b2f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-588ddfd.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-588ddfd.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9ea28ab8d673366f830b45f6a306bd739bb6f89e available at: https://gist.github.com/9a74abfb6c64450646c71046862e7b2f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-9ea28ab.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10085-9ea28ab.dmg

**Comment by Nex-Pro** [comment]
### PASS

**Pull request merged**